cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/query/QueryResult.java

```diff
@@ -34,6 +35,18 @@
  */
 public class QueryResult implements Iterable<DocumentRevision> {
 
+    /**
+     * This unchecked exception is used to wrap another exception so
+     * we can pass meaningful exception traces over the {@link #iterator()}
+     * boundary as we can't change the signature of {@link #iterator} to
+     * add a checked exception.
+     */
+    public class UncheckedDocumentException extends RuntimeException {
+        public UncheckedDocumentException(Exception causedBy){
+            super(causedBy);
+        }
+    }
```

Comment:
This is a bad choice of name. A class enclosed within QueryResult shouldn't be called something as generic as `*DocumentException`.

Given we need to wrap the exception to maintain compatibility with the existing interface, we should probably go for something that's more generic across the querying code. I'd go for something like `com.cloudant.sync.query.QueryException extends RuntimeException`. That name also implies putting it in it's own class.

We can add a TODO to make this exception checked in 2.0 if we want.